src/web/middleware.rs

```diff
@@ -1,151 +1,273 @@
-            .map_err(Into::into);
-        Ok(Started::Future(Box::new(fut)))
+impl DbTransaction {
+    pub fn new() -> Self {
+        DbTransaction::default()
     }
+}
 
-    fn response(&self, req: &HttpRequest<ServerState>, resp: HttpResponse) -> Result<Response> {
-        if let Some((db, in_transaction)) = req.extensions().get::<(Box<dyn Db>, bool)>() {
-            if *in_transaction {
-                let fut = match resp.error() {
-                    None => db.commit(),
-                    Some(_) => db.rollback(),
-                };
-                let fut = fut.and_then(|_| Ok(resp)).map_err(Into::into);
-                return Ok(Response::Future(Box::new(fut)));
+impl Default for DbTransaction {
+    fn default() -> Self {
+        Self
+    }
+}
+
+impl<S, B> Transform<S> for DbTransaction
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S: 'static,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = DbTransactionMiddleware<S>;
+    type Future = FutureResult<Self::Transform, Self::InitError>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        future::ok(DbTransactionMiddleware {
+            service: Rc::new(RefCell::new(service)),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct DbTransactionMiddleware<S> {
+    service: Rc<RefCell<S>>,
+}
+
+impl<S, B> Service for DbTransactionMiddleware<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S: 'static,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = Box<Future<Item = Self::Response, Error = Self::Error>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.service.poll_ready()
+    }
+
+    fn call(&mut self, sreq: ServiceRequest) -> Self::Future {
+        // `into_parts()` consumes the service request.
+        let method = sreq.method().clone();
+        let col_result = CollectionParam::extrude(&sreq.uri(), &mut sreq.extensions_mut());
+        let collection = match col_result {
+            Ok(v) => v,
+            Err(e) => {
+                dbg!("⚠️ CollectionParam err: {:?}", e);
+                return Box::new(future::ok(
+                    sreq.into_response(
+                        HttpResponse::InternalServerError()
+                            .content_type("application/json")
+                            .body("Err: invalid collection".to_owned())
+                            .into_body(),
+                    ),
+                ));
+            }
+        };
+        let ci = &sreq.connection_info().clone();
+        let headers = &sreq.headers();
+        let auth = match headers.get("authorization") {
+            Some(a) => a.to_str().unwrap(),
```

Comment:
There's a few `unwrap()`s in this file that need cleanup, I'll log an issue to fix these after this merges